File: DynamicProgramming/change/dinamic_memo.py
from functools import lru_cache

INF = float('inf')

def main():
    moedas = [1, 3, 4]
    valor = 6

    resultado = solve(valor, tuple(moedas))  
    print(resultado)

@lru_cache(maxsize=None)  # memoization ilimitada
def solve(x, moedas):
    if x < 0:
        return INF
    if x == 0:
        return 0
    best = INF
    for c in moedas:
        best = min(best, solve(x - c, moedas) + 1)
    return best

if __name__ == "__main__":
    main()

File: DynamicProgramming/change/dinamic_memo.cpp
#include <iostream>
#include <vector>
#include <algorithm>

const int INF = 1e9;

int solve(int x, const std::vector<int>& moedas, std::vector<int>& memo) {
    if (x < 0) return INF;
    if (x == 0) return 0;
    if (memo[x] != -1) return memo[x]; // já calculado

    int best = INF;
    for (auto c : moedas) {
        best = std::min(best, solve(x - c, moedas, memo) + 1);
    }
    
    memo[x] = best;
    return best;
}

int main() {
    std::vector<int> moedas = {1, 3, 4};
    int valor = 6;

    std::vector<int> memo(valor + 1, -1); // -1 = ainda não calculado

    int resultado = solve(valor, moedas, memo);
    std::cout << resultado << std::endl;

    return 0;
}